src/encoded/types/cart.py

```diff
@@ -99,47 +88,46 @@ def get_or_create_cart_by_user(context, request):
     user = request.registry[COLLECTIONS]['user'].get(userid)
     if not user:
         raise HTTPBadRequest()
-    carts = _get_carts_by_user(request, userid, ['disabled', 'deleted '])
+    carts = get_cart_objects_by_user(request, userid, blocked_statuses)
+    return carts, user
+
+
+@view_config(context=Cart.Collection, request_method='GET', permission='save-carts', name='get-cart')
+def get_or_create_cart_by_user(context, request):
+    carts, user = _get_user_carts(request, ['disabled', 'deleted '])
     if not carts:
         cart = _create_cart(request, user)
+        cart_atids = None
+    else:
+        cart_atids = [cart['@id'] for cart in carts]
     request.response.status = 200
     return {
         'status': 'success',
         '@type': ['result'],
-        '@graph': carts or [cart]
+        '@graph': cart_atids or [cart]
     }
 
 
 @view_config(context=Cart.Collection, request_method='PUT', permission='save-carts', name='put-cart')
 def create_cart_by_user(context, request):
-    userid = [
-        p.replace('userid.', '')
-        for p in request.effective_principals
-        if p.startswith('userid.')
-    ]
-    if not userid:
-        raise HTTPBadRequest()
-    else:
-        userid = userid[0]
-    user = request.registry[COLLECTIONS]['user'].get(userid)
-    if not user:
-        raise HTTPBadRequest()
-    cart_name = None
-    user_carts = get_cart_objects_by_user(request, userid)
-    visible_user_carts = filter_carts_to_visible(request, user_carts)
-    countable_carts = [
-        cart
-        for cart in visible_user_carts
-        if cart['status'] != 'disabled'
-    ]
-    cart_atids = [cart['@id'] for cart in visible_user_carts] if visible_user_carts else []
+    user_carts, user = _get_user_carts(request, ['deleted'] if not 'group.admin' in request.effective_principals else [])
+    cart_atids = [cart['@id'] for cart in user_carts]
     cart_status = request.json.get('status', None)
     cart_name = request.json.get('name', '').strip()
     # User writing a new cart; check for cart overflow and naming conflicts
+    if cart_status != 'disabled':
+        countable_carts = [
+            cart
+            for cart in user_carts
+            if cart['status'] != 'disabled' and cart['status'] != 'deleted'
+        ]
+    else:
+        # Creating an auto-save cart, so ignore existing carts
+        countable_carts = []
     if len(countable_carts) >= CART_USER_MAX:
         msg = 'Users cannot have more than {} carts'.format(CART_USER_MAX)
         raise HTTPBadRequest(explanation=msg)
-    elif next((cart for cart in visible_user_carts if (cart['status'] != 'deleted' and
+    elif next((cart for cart in user_carts if (cart['status'] != 'deleted' and
```

Comment:
@caseylitton points out you don't have to use elif here since previous if raises error. The `any` and `all` keywords are pretty standard and imply that you are checking condition of items in list, which is what you are doing here. The `next` keyword implies that you are yielding a value in order to do something with it, which is confusing in this case because you aren't doing anything with the item. It can work well if you want to return first item from list if it exists, but aren't sure if it does.

Reply:
Changed and tested, and it seems to work the same as before. I agree that makes more sense.